pep-0517.txt

```diff
@@ -636,14 +634,51 @@ build backend::
                 return archive_entry
         return None
 
-    def build_sdist(sdist_dir, config_settings):
-        sdist_subdir = "mypackage-0.1"
-        sdist_path = pathlib.Path(sdist_dir) / (sdist_subdir + ".tar.gz")
+    def _make_sdist(sdist_dir):
+        """Make an sdist and return both the Python object and its filename"""
+        sdist_path = pathlib.Path(sdist_dir) / SDIST_FILENAME
         sdist = tarfile.open(sdist_path, "w:gz", format=tarfile.PAX_FORMAT)
         # Tar up the whole directory, minus hidden and special files
-        sdist.add(os.getcwd(), arcname=sdist_subdir,
+        sdist.add(os.getcwd(), arcname=SDIST_NAME,
                   filter=_exclude_hidden_and_special_files)
-        return sdist_subdir + ".tar.gz"
+        return sdist, SDIST_FILENAME
+
+    def build_sdist(sdist_dir, config_settings):
+        """PEP 517 sdist creation hook"""
+        sdist, sdist_filename = _make_sdist(sdist_dir)
+        return sdist_filename
+
+    #################
+    # wheel creation
+    #################
+
+    def get_requires_for_build_wheel(config_settings):
+        """PEP 517 wheel building dependency definition hook"""
+        # As a simple static requirement, this could also just be
+        # listed in the project's build system dependencies instead
+        return ["wheel"]
+
+    def _prepare_out_of_tree_build(build_directory):
+        """Prepare out-of-tree build by way of unpacking the sdist"""
+        sdist, sdist_filename = _make_sdist(build_directory)
+        os.chdir(build_directory)
+        sdist.extractall()
+        os.remove(sdist_filename)
+        os.chdir(SDIST_NAME)
+
+    def build_wheel(wheel_directory, build_directory=None,
+                    metadata_directory=None, config_settings=None):
+        """PEP 517 wheel creation hook"""
+        # First check if the frontend has requested an out-of-tree build
+        out_of_tree_build = (build_directory is not None and
+                             not os.path.samefile(build_directory, os.getcwd())
```

Comment:
Do we want to make a note about this case (`build_directory==source_directory`) in the spec? I would have written the code assuming that if build_directory is not None, we're doing an out of tree build. I don't mind whether we say that it can be the same as the source directory or that it should be different, but we should probably make it explicit either way.